### PR TITLE
Test app

### PR DESCRIPTION
**Pull request by mhsmith**
At this stage I haven't yet set the project up to use separate tests per platform. I expect this would be done in pyproject.toml by making the `windows` section contain the line `sources = ["../winforms/tests"]`, and so on.

So far, I've added example Winforms tests for:
* Native widget existence and type
* A property
* A handler

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by mhsmith** [comment]
The whole test suite is currently being run synchronously on the main thread. This hasn't caused any problems so far, but I'm sure it will soon, so that's the next thing I'll look at. 

We talked about the possibility of making Toga itself thread-safe, so that every method in the Toga API would run itself on the main thread if necessary. But if we did that, and then ran the tests on a background thread, it wouldn’t be representative of the common single-threaded case in a real app, because we would be executing each Toga API call on the main thread separately, with the possibility of other things happening in between.

So I think the best approach is to run the pytest main loop on a background thread, but arrange for each test function and its supporting fixtures to be run on the main thread. The pytest thread would block until the function completed, and any exception would be re-raised on the pytest thread so it could be reported.

Fortunately It looks like pytest has a comprehensive hook system which should make this fairly straightforward.

File: test/src/toga_test/app.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+        self.main_window = toga.MainWindow(title=self.formal_name)
+        self.main_window.content = self.main_box
+        self.main_window.show()
+
+        self.add_background_task(test)
+
+
+def test(app):
+    pytest.main(["-vv", "--color=no"] + tests.__path__)
```
Comment (mhsmith) [inline test/src/toga_test/app.py:23]: This probably won't work on Android because its Python modules don't exist as separate files on the filesystem. As with the threading issue, it looks like this can be dealt with using pytest's hook system.

File: test/src/tests/widgets/test_label.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    native_box = main_box._impl.native
+    assert native_box.Controls.Count == 1
+    # TODO: assert native_box.Controls[0] is native
+    assert isinstance(native_box.Controls[0], WinForms.Label)
+
+
+def test_text(widget, native):
+    for text in TEXTS:
+        set_get(widget, "text", text)
+        assert native.Text == text
```
Comment (mhsmith) [inline test/src/tests/widgets/test_label.py:29]: This is a test of setting a property *after* construction. 

As long as the core tests verify that setting a property *during* construction causes the same sequence of dummy backend calls as setting it *after* construction, it shouldn't be necessary to have per-backend tests for the contructor arguments. But it doesn't look like the core tests do that at the moment.

File: test/src/tests/widgets/test_label.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    main_box.add(label)
+    yield label
+    main_box.remove(label)
+
+
+def test_native(widget, native, main_box):
+    native_box = main_box._impl.native
+    assert native_box.Controls.Count == 1
+    # TODO: assert native_box.Controls[0] is native
+    assert isinstance(native_box.Controls[0], WinForms.Label)
```
Comment (mhsmith) [inline test/src/tests/widgets/test_label.py:23]: The `is` check doesn't currently work, so maybe pythonnet doesn't enforce a 1-1 relationship between .NET objects and their Python proxies. I couldn't find any information about this in a quick search. @freakboy3742: any idea?

File: test/src/toga_test/app.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        global app
+        app = self
+
+        self.main_box = toga.Box()
+
+        self.main_window = toga.MainWindow(title=self.formal_name)
+        self.main_window.content = self.main_box
+        self.main_window.show()
+
+        self.add_background_task(test)
```
Comment (freakboy3742) [inline test/src/toga_test/app.py:19]: Not sure how well this approach will work in the general case - `test()` is a blocking method, so that means the main event loop won't be actively running while the tests are running. That will definitely interfere with any display updates that are the result of tests.

**Review by freakboy3742 — COMMENTED** [review]
A couple of comments inline

File: test/src/toga_test/app.py (hunk 4 above)
Comment (mhsmith) [reply test/src/toga_test/app.py:19]: See [comment above](https://github.com/beeware/toga/pull/1687#issuecomment-1315914548): this is the next thing I'll look at.

File: test/src/tests/widgets/test_label.py (hunk 3 above)
Comment (mhsmith) [reply test/src/tests/widgets/test_label.py:23]: It looks like [`Object.ReferenceEquals`](https://learn.microsoft.com/en-us/dotnet/api/system.object.referenceequals?view=net-7.0) does what we need.

File: test/src/toga_test/app.py (hunk 1 above)
Comment (mhsmith) [reply test/src/toga_test/app.py]: I've added a chaquopy_extract_package function to the bottom of app.py which extracts the given package into individual files on the filesystem. This could be used in the testbed app as well.

A cleaner solution to this might be to resurrect a Chaquopy setting called [`extractPackages`](https://chaquo.com/chaquopy/doc/6.2/android.html#resource-files). This used to be required for any packages that expected to find its non-Python files on the filesystem, but it turned out there were so many packages like that, that I just made Chaquopy ignore the setting and extract data files automatically, but not Python files. 

That works for 99% of packages, but occasionally I get a user report of some code which expects Python files to exist on the filesystem as well.  So we could re-enable the `extractPackages` setting and make it mean "extract *all* files in this package, including Python files".

Comment (mhsmith) [reply test/src/toga_test/app.py]: In cases like this, the code probably expects to find files named `.py`, not `.pyc`. So for anything listed in `extractPackages`, Chaquopy should preserve the original .py files, which it normally omits if it succeeded in compiling them to .pyc format during the build.

For the moment, with pytest, we'll need to work around this by using the [`pyc`](https://chaquo.com/chaquopy/doc/current/android.html#android-bytecode) setting to disable build-time compilation, So the `python` section in app/build.gradle would look like this:
```
        python {
            version "3.8"
            pip {
                install "-r", "requirements.txt"
            }
            pyc {
                src false
            }
        }
```

To enable this, and the future use of `extractPackages`, and other user customizations, how about adding a pyproject.toml setting `build_gradle_extra_content`, that works the same way as [`dockerfile_extra_content`](https://github.com/beeware/briefcase-linux-appimage-template/pull/22)?

Gradle's brace syntax is really just shorthand for dotted notation. For example, the above block is equivalent to this:
```
        python {
            version "3.8"
            pip {
                install "-r", "requirements.txt"
            }
        }

        python {
            pyc {
                src false
            }
        }
```

Or this:
```
        python {
            version "3.8"
            pip {
                install "-r", "requirements.txt"
            }
        }

        python.pyc.src false
```

So adding the `build_gradle_extra_content` at the end of the file should be fine: it would still be able to customize any of the settings in the blocks above.

Comment (freakboy3742) [reply test/src/toga_test/app.py]: Exposing Gradle customisation definitely seems like a reasonable thing to do - no matter what we put in the template, I'm sure *someone* will have other things they want to do. If it's as simple as adding strings of dotted paths to the bottom of a file, then it seems like a no-brainer.

I'd also be in favour of adding the `extractPackages` setting; it's got immediate utility for test suites, and I have to imagine there's a set of weird-edge case packages that will actually want to have their source files on disk for Reasons (tm).

File: test/pyproject.toml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+[tool.briefcase.app.toga-test.linux.flatpak]
+flatpak_runtime = 'org.gnome.Platform'
+flatpak_runtime_version = '42'
+flatpak_sdk = 'org.gnome.Sdk'
+
+[tool.briefcase.app.toga-test.windows]
+sources = [
+    '../winforms/test_probes',
+]
```
Comment (mhsmith) [inline test/pyproject.toml:60]: Once we get the existing backend tests working within the test app, and once https://github.com/beeware/briefcase/issues/975 is fixed, this can be merged into the existing directory `../winforms/tests`.

File: test/src/tests/widgets/test_label.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+        set_get(widget, "text", text)
+        assert probe.text == text
+
+
+async def test_color(widget, probe):
+    for color in COLORS:
+        widget.style.color = color
+        for component in ["r", "g", "b"]:
+            assert getattr(probe.color, component) == getattr(color, component)
+        assert probe.color.a == approx(color.a, abs=(1 / 256))
```
Comment (mhsmith) [inline test/src/tests/widgets/test_label.py:25]: An example of a property where the value coming out is only approximately equal to the value going in, because of the way it's represented in the native API.

File: test/src/tests/widgets/test_label.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+async def test_multiline(widget, probe):
+    def make_lines(n):
+        return "\n".join(f"line{i}" for i in range(n))
+
+    widget.text = make_lines(1)
+    line_height = probe.height
+
+    widget.text = make_lines(2)
+    assert probe.height == approx(line_height * 2, rel=0.1)
+    line_spacing = probe.height - (line_height * 2)
```
Comment (mhsmith) [inline test/src/tests/widgets/test_label.py:37]: This test currently fails on Windows, probably because of #1289, and maybe also because we need an `await` to allow the size change to take effect, as we will on Android.

File: winforms/test_probes/widgets/probe_slider.py [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+    native_class = System.Windows.Forms.TrackBar
+
+    @property
+    def position(self):
+        min, max = self._range
+        return (self.native.Value - min) / (max - min)
+
+    def change(self, position):
+        min, max = self._range
+        self.native.Value = min + round(position * (max - min))
```
Comment (mhsmith) [inline winforms/test_probes/widgets/probe_slider.py:16]: Methods to trigger events are named the same as the corresponding event handler.

**Review by mhsmith — COMMENTED** [review]
@freakboy3742: This is still incomplete, but I'd be interested to know your thoughts on the general approach.

File: test/src/tests/widgets/test_label.py (hunk 6 above)
Comment (freakboy3742) [reply test/src/tests/widgets/test_label.py:25]: That sounds like something we might want to pull out into a utility library; color comparisons are going to be common.

File: winforms/test_probes/widgets/probe_slider.py (hunk 8 above)
Comment (freakboy3742) [reply winforms/test_probes/widgets/probe_slider.py:16]: Is that purely a convention, or is there a functional consequence?

Comment (mhsmith) [reply winforms/test_probes/widgets/probe_slider.py:16]: Just a convention.

**Review by freakboy3742 — COMMENTED** [review]
tl;dr - I like what I'm seeing. Some of the rough edges are definitely due to not being in "briefcase run --test" format yet, plus some of the complications you've flagged about shared sources. I've got a couple of questions about the usage of fixtures, plus a couple of cosmetic quirks, but on the whole, this definitely looks like it's heading in the right direction.

A couple of "next steps" features I'd like to see:
* Geometry tests. Proving that a widget expands to the expected size in a given layout
* Tests of a widget like NumericalInput where there's a range of inputs, and the input isn't necessarily the same as the output
* Tests of a widget that has focus-driven behavior (e.g., text input doing input validation)
* A "wait for the GUI to do it's thing" await call; you've flagged the possible need for one of these, but it strikes me it could be a bit of a "draw the rest of the owl" situation

File: test/src/tests/widgets/test_button.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
+from unittest.mock import Mock
+
+from pytest import fixture
+
+import toga
+
+
+@fixture
+async def new_widget():
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/src/tests/widgets/test_button.py:9]: > Do all fixtures have to be async?

Yes, if they call any Toga APIs, because making them async is what causes them to run on the main thread.

> Is a fixture the right model here - especially for the "broader layout"? Might it not be easier to have a `simple_layout(widget)` which puts the test widget into a simple box layout?

I'm not quite sure what you're suggesting, but I'll rename the fixtures so it's clearer what each one is actually doing.

File: test/src/tests/widgets/test_label.py (hunk 7 above)
Comment (mhsmith) [reply test/src/tests/widgets/test_label.py:37]: In order to keep this PR manageable, I won't include any fixes for bugs in Toga itself. If I come across any more while adding tests, I'll just skip the test with a comment.

File: test/src/tests/widgets/test_label.py (hunk 6 above)
Comment (mhsmith) [reply test/src/tests/widgets/test_label.py:25]: Good point, I'll do that.

**Comment by mhsmith** [comment]
I've now added an initial set of tests for Label, Button and Slider, with probe implementations for Android and Windows.

Running on Android will require https://github.com/beeware/briefcase-android-gradle-template/pull/58.

This is not intended to give complete coverage of any widget: that can be done in later PRs. In particular:
* Tests for features not currently supported by a backend are skipped by calling `pytest.skip` in the probe. 
* Tests which should be supported but currently fail are skipped with `pytest.mark.skip`.

File: android/test_probes/widgets/probe_button.py [hunk 10]
```diff
@@ -0,0 +1,8 @@
+from java import jclass
```
Comment (freakboy3742) [inline android/test_probes/widgets/probe_button.py:1]: A slightly bikeshed comment: but how would you feel about:

* Renaming `test_probes/probe_button.py::ButtonProbe` to `probes/button_probe.py::ButtonProbe`
* Adding `from .button_probe import ButtonProbe` to `probes/__init__.py`.

That way:
1. The namespace collision issue with copying test code isn't an problem any more
2. There's consistency of naming between `button_probe.py` and `ButtonProbe` (FWIW, ˆ still prefer for `probes/button.py`, even if there is multiple button.py files; but I can live with a longer name if it's at least consistent)
3. It's clear about isolating "test instrumentation" from "test code"
4. The usage pattern in tests will be `from probes import ButtonProbe`.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Generally looking good; a variety of questions and comments inline.

File: test/pyproject.toml [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+icon = "src/toga_test/resources/toga_test"
+sources = [
+    'src/toga_test',
+]
+test_sources = [
+    'tests',
+]
+requires = [
+    '../core',
+    '../dummy',
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/pyproject.toml:23]: I added it with the idea that we would run the existing test_implementation tests within the test app, but that won't be happening in this PR, so I'll remove it for now.

File: test/src/toga_test/app.py [hunk 12]
```diff
@@ -0,0 +1,18 @@
+import toga
+
+
+class TogaTest(toga.App):
+    def startup(self):
+        # For the "app" fixture.
+        global app
+        app = self
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/src/toga_test/app.py:8]: Ah, I'd forgotten about `toga.App.app`. Fixed to use that instead.

File: test/tests/conftest.py [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+@fixture(scope="session")
+def main_box(app):
+    return app.main_box
+
+
+# Controls the event loop used by pytest-asyncio.
+@fixture(scope="session")
+def event_loop(app):
+    return ProxyEventLoop(app._impl.loop)
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/conftest.py:23]: I can see the advantages of that, but I usually prefer to go from high to low level, so you can understand the big picture first, and then either skip over the details if they're not important yet, or read them with an understanding of the context in which they're used. But as long as the approach is consistent within any given file, I think it doesn't matter that much.

File: test/tests/utils.py [hunk 14]
```diff
@@ -0,0 +1,32 @@
+from pytest import approx
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/utils.py]: OK, I've split `tests.utils` into `data` and `assertions`, and I've also renamed:
* `tests.widgets.common` to `tests.widgets.properties`
* `test_probes.utils` to `test_probes.widgets.properties`

File: test/tests/widgets/test_button.py [hunk 15]
```diff
@@ -0,0 +1,22 @@
...
+
+from .common import test_background_color, test_color, test_text  # noqa: F401
+
+
+@fixture
+async def widget():
+    return toga.Button("")
+
+
+async def test_press(widget, probe):
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/widgets/test_button.py:19]: OK, let's discuss this tomorrow.

File: test/tests/widgets/common.py [hunk 16]
```diff
@@ -0,0 +1,19 @@
+from ..utils import COLORS, TEXTS, assert_color, assert_set_get
+
+
+async def test_text(widget, probe):
+    for text in TEXTS:
```
[the remark this answers is not in the document]
Comment (freakboy3742) [reply test/tests/widgets/common.py]: Ah - that makes sense; although, it also means that the specific transitions from TEXTS[0]->TEXTS[1]->... are being tested as well. That's a good thing, I suppose; but we need to be clear when writing up the sample data whether the it's actual literal value that is of interest, or the transition from one value to another.

File: test/tests/conftest.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply test/tests/conftest.py:23]: Fair enough; both approaches have decent reasons. At some point we might want to consider a project-wide style guide, but that's an elephant we don't need to eat at this particular sitting.

File: test/tests/widgets/test_button.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply test/tests/widgets/test_button.py:19]: After discussion, we came to the conclusion that:

* Yes, this is potentially confusing for people not particularly familiar with Pytest; however, it's also very much learnable knowledge, so anyone with moderate experience with Pytest should be familiar with the trick being used here
* In the specific case of these tests, we have a need for "a widget and a probe" for every test. That's likely to be a common property of almost all the individual widget tests - enough that having a single simple abstraction makes sense.
* If we have other multi-widget or complex-layout requirements, we can put them in separate testing packages with isolated conftest modules that won't inherit this specific widget + probe definitions.

[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/widgets/test_button.py:19]: I think for someone who understands the basics of pytest, the only unusual thing going on here is that we have a `probe` fixture in a conftest file depending on a `widget` fixture in a test module. This is less common than the other direction, but conceptually it's no different from a method in a base class calling an abstract method which must be implemented by a subclass.

So maybe it would be clearer if the conftest file actually contained an abstract `widget` fixture which raises NotImplementedError, with a comment explaining how the test modules are supposed to override it. That way you could read the conftest file in isolation without being confused about where the `widget` was coming from. And if you forgot to define `widget` in a test module, then instead of pytest's generic "missing fixture" error, you would get a NotImplementedError leading you directly to the relevant information.

Anyway, I'll revisit this along with the comment above about the probe constructor.

File: android/test_probes/widgets/probe_button.py (hunk 10 above)
Comment (mhsmith) [reply android/test_probes/widgets/probe_button.py:1]: Now that we have a few examples in place, I think you were right that it's better to keep the common test code and the backend-specific code in separate namespaces. So since we already have `tests`, I've gone with `tests_backend`, to mirror the relationship between `toga` and `toga_cocoa` etc., and because there will probably be things other than probes in there soon enough.

And I've removed the `probe_` prefix on the module names as you suggest.

File: android/test_probes/widgets/probe_base.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
+from pytest import skip
+
+
+class SimpleProbe:
+    def __init__(self, main_box, widget):
+        native_box = main_box._impl.native
+        assert native_box.getChildCount() == 1
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply android/test_probes/widgets/probe_base.py:7]: Good point: I've reduced the Probe constructors to just be "find this widget inside this container", and any more specific layout assertions can wait until we write the layout tests.

File: test/tests/widgets/conftest.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
+from importlib import import_module
+
+from pytest import fixture, skip
+
+
+@fixture
+async def simple_layout(main_box, widget):
+    main_box.add(widget)
+    yield
+    main_box.remove(widget)
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/widgets/conftest.py:10]: OK, I've made it create a new box for each test.

File: android/test_probes/widgets/probe_base.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (mhsmith) [reply android/test_probes/widgets/probe_base.py:7]: And I've separated out the function that instantiates the probe so it can be called from more complex layouts.

File: test/tests/utils.py [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+
+# TODO: include None
+components = [0, 1, 2, 128, 253, 254, 255]
+COLORS = [
+    rgba(r, g, b, a)
+    for r in components
+    for g in components
+    for b in components
+    for a in [0.0, 0.01, 0.1, 0.5, 0.9, 0.99, 1.0]
+]
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/utils.py]: Haha, I hadn't calculated how many colors that would multiply up to. I've cut it down to a shorter list.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
A couple more questions/queries inline; some of them can probably be punted until we're actually building tests for widgets, but a couple are more fundamental.

One more bikeshedding thing around project layout; this PR adds a `test` (singular) folder that adds a `test_app` project & app that has a `tests` folder inside it. 

I've historically leant towards using `tests` for folders that contain multiple tests, as the Python standard library has a `test` top level module. Although that *shouldn't* cause problems, one rare occasions where you're trying to sort out `PYTHONPATH` issues related to the test suite, it's *much* easier to debug when there's no top-level name clash.

In this case, that particular problem isn't an issue (other than as a matter of consistency), as it's not just a folder of tests; it's an app that is needed to run a suite of tests. The Python-support-testbed has established a precedent for calling apps of this type "testbed", and I think that name makes as much sense here as it does in the support package testbed - it's a block of scaffolding that lets you deploy the actual tests that you want to perform.

So, I wonder if we should continue that convention here:

* rename the top level `test` folder to `testbed`.
* rename the `toga_test` app to:
   * app/module name `testbed`
   * formal name "Toga Testbed"
   * bundle ID `org.beeware.toga.testbed`

That would maintain the naming consistency between projects; but also makes it clear that the "tests" folder isn't just a suite of tests you can run `pytest` over; it needs to be deployed.

File: test/src/toga_test/app.py [hunk 20]
```diff
@@ -0,0 +1,12 @@
...
+
+class TogaTest(toga.App):
+    def startup(self):
+        self.main_window = toga.MainWindow(title=self.formal_name)
+        self.main_window.content = toga.Box()
+        self.main_window.show()
+
+
+def main():
+    return TogaTest(app_name="toga_test")
```
Comment (freakboy3742) [inline test/src/toga_test/app.py:12]: This is odd... why is app_name needed here? It should be loading this from the `dist-info`...

Comment (freakboy3742) [reply test/src/toga_test/app.py:12]: Oh - I get it now - it's because the entry point app is `tests.toga_test`, which doesn't have a corresponding dist-info; passing `app_name` is enough to point it at the right dist-info file.

File: test/tests/toga_test.py [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+            "--color=no",  # Color codes mess up the output on MSYS2, even with `winpty`.
+            # Run all async tests and fixtures using pytest-asyncio.
+            "--asyncio-mode=auto",
+            # Override the cache directory to be somewhere known writable
+            "-o",
+            f"cache_dir={tempfile.gettempdir()}/.pytest_cache",
+        ]
+        + tests.__path__
+    )
+    app.add_background_task(lambda app: app.exit())
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/toga_test.py:33]: `run_tests` runs on a background thread, so it's not safe for it to make direct calls to Toga APIs. The previous `add_background_task` call was just to queue the `Thread.start` call, to make sure the `run_tests` thread doesn't start until after the UI is ready.

File: android/tests_backend/widgets/slider.py [hunk 22]
```diff
@@ -0,0 +1,24 @@
...
+    @property
+    def position(self):
+        return (self.native.getProgress() - self._min) / (self._max - self._min)
+
+    def change(self, position):
+        self.native.setProgress(self._min + round(position * (self._max - self._min)))
+
+    @property
+    def _min(self):
+        return 0 if (Build.VERSION.SDK_INT < 26) else self.native.getMin()
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply android/tests_backend/widgets/slider.py:20]: It won't make any difference here, because Toga doesn't currently use the minimum value property: I just added this code to the probe in case that ever changes. 

But as a general principle yes, I think it would be a good idea to run Android tests against both the minimum and maximum API levels we support, and that's what I always do before a Chaquopy release. 

> If I'm understanding correctly, we would only need to _build_ it once, but we'd need to use 2 simulators?

That's correct.

File: test/tests/widgets/conftest.py [hunk 23]
```diff
@@ -0,0 +1,19 @@
...
+async def widget():
+    raise NotImplementedError("test modules must define a `widget` fixture")
+
+
+@fixture
+async def probe(main_window, widget):
+    box = toga.Box(children=[widget])
+    main_window.content = box
+    yield get_probe(widget, box)
+    # TODO: Window.content doesn't currently accept None.
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/widgets/conftest.py:18]: OK, I'll remove the comment.

File: test/src/toga_test/app.py [hunk 24]
```diff
@@ -0,0 +1,12 @@
+import toga
+
+
+class TogaTest(toga.App):
+    def startup(self):
+        self.main_window = toga.MainWindow(title=self.formal_name)
+        self.main_window.content = toga.Box()
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/src/toga_test/app.py]: > It might be worth adding a placeholder text label here letting the user know that this is a test suite, and they should probably be using `--test`

Fixed by merging https://github.com/mhsmith/toga/pull/1.

> We could possibly use this as a HTML-style [ACID test](https://www.acidtests.org) - a layout that has a very specific visual appearance that lets us do an easy sanity check of visual properties that can't be easily tested.

Do you mean using the the placeholder text for this, or something more complex? Either way, I think the best plan is to pay attention to what bug reports we get for each widget after we've fully tested it, and use that as guidance for what this approach is covering well or not so well.

**Comment by mhsmith** [comment]
> So, I wonder if we should continue that convention here:
> 
> * rename the top level `test` folder to `testbed`.
> * rename the `toga_test` app to:
>   
>   * app/module name `testbed`
>   * formal name "Toga Testbed"
>   * bundle ID `org.beeware.toga.testbed`

That all sounds reasonable: fixed by merging https://github.com/mhsmith/toga/pull/1.

**Comment by mhsmith** [comment]
I've had enough of these random Codecov failures, and we're not really using it, so I've just removed it.

File: test/tests/toga_test.py [hunk 25]
```diff
@@ -0,0 +1,55 @@
...
+        if finder.isdir(zip_path):
+            chaquopy_extract_dir(finder, zip_path)
+        else:
+            finder.extract_if_changed(zip_path)
+
+
+if __name__ == "__main__":
+    app = main()
+    thread = Thread(target=partial(run_tests, app))
+    app.add_background_task(lambda app: thread.start())
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/toga_test.py]: OK, I'll update this.

File: test/tests/widgets/test_slider.py [hunk 26]
```diff
@@ -0,0 +1,67 @@
...
+async def widget(on_change):
+    return toga.Slider(on_change=on_change)
+
+
+@fixture
+def on_change():
+    return Mock()
+
+
+@mark.skip("Fails on Android: initial value is 0.0")
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply test/tests/widgets/test_slider.py]: Turns out they actually fail on both platforms, but for different reasons, so I've updated the decorators to document that.

File: android/tests_backend/widgets/base.py [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+class SimpleProbe:
+    def __init__(self, widget, container):
+        for i in range(container._impl.native.getChildCount()):
+            child = container._impl.native.getChildAt(i)
+            if child is widget._impl.native:
+                self.native = child
+                assert isinstance(self.native, self.native_class)
+                break
+        else:
+            raise ValueError(f"cannot find {widget} in {container}")
```
[the remark this answers is not in the document]
Comment (mhsmith) [reply android/tests_backend/widgets/base.py:13]: The idea behind "simple" was that it covers the case where there's a 1-1 relationship between Toga widgets and native widgets. But I know you've mentioned that some Toga widgets on some platforms might be implemented as multiple native widgets. Are there any widgets like this already? If so, their probes couldn't inherit from this base class, because there wouldn't just be a single `self.native` property to implement all the methods against.

[the remark this answers is not in the document]
Comment (mhsmith) [reply android/tests_backend/widgets/base.py:13]: > However, this test is a specific test of the type of layout it's been added to.

It's no longer making any assumptions except that the widget is somewhere inside the container. What kind of layout would that not work with?

The idea behind this code is to verify that adding a Toga widget to a Toga container causes a native widget of the correct type to be added to the native container. Or in the case of a Toga widget which produced multiple native widgets, the counterpart to this code would also need to verify that *all* the native widgets had been added to the layout, and in the correct configuration.

> Pulling this block of code out as an `assert_has_child(container, widget)` assertion in the main test app (and invoked in the probe setup) would also mean we're being explicit about what has been asserted.

Good point, and that would also allow us to call the code multiple times on the same widget, to verify that a widget can be moved from one container to another. So I'll move it out of the constructor.

**Comment by mhsmith** [comment]
Feel free to push directly to this branch if you know what's happening with the Linux CI failure.

**Comment by mhsmith** [comment]
I guess to prevent Codecov leaving comments it would needs to be disabled in the repository settings as well.

File: test/tests/toga_test.py (hunk 21 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply test/tests/toga_test.py]: Ah - of course. 

/me returns the "days since Russell got confused about threads" sign back to 0.

File: test/src/toga_test/app.py (hunk 24 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply test/src/toga_test/app.py]: I was thinking something more complex. I don't have anything particular in mind; I was just thinking that it was real estate that we could use for a purpose, and a "visual layout that is known to be problematic" seemed like a useful purpose. However, as you say - we can use bug reports to guide this.

File: android/tests_backend/widgets/base.py (hunk 27 above)
[the remark this answers is not in the document]
Comment (freakboy3742) [reply android/tests_backend/widgets/base.py]: > Are there any widgets like this already? If so, their probes couldn't inherit from this base class, because there wouldn't just be a single `_impl.native` property to implement all the methods against.

There are several widgets that do this. On macOS, the scrolling behavior for MultilineText is independent of the widget itself, so the "native" is the scroller, and there's a secondary widget that is the text view. Switch is another example; this time, on macOS it's a single widget, but on GTK it's a Label and a Switch, wrapped up in a Box.

**Comment by freakboy3742** [comment]
I've uninstalled CodeCov from the project; agreed that it has outstayed its welcome.

**Comment by mhsmith** [comment]
The fix for https://github.com/beeware/briefcase/pull/992 works great, so the Linux app is building in CI now, but when it tries to run it simply segfaults. 😦

**Comment by mhsmith** [comment]
Yet it works fine on my Linux machine. 

The only thing I can think of is maybe we should be running it through xvfb-run, as the backend unit tests do in tox.ini. Though if it was that, I would expect it to give a more useful error message.

**Comment by mhsmith** [comment]
> The only thing I can think of is maybe we should be running it through xvfb-run, as the backend unit tests do in tox.ini. Though if it was that, I would expect it to give a more useful error message.

Well, it turns out it was that. And I can reproduce the crash on my own Linux machine simply by setting the DISPLAY environment variable to empty to prevent the app from contacting the X server. That's a pretty terrible user experience, but I don't know which component is to blame.

**Comment by mhsmith** [comment]
Remember when the support testbed app was reporting tests taking negative time? Well, if you look at [this line](https://github.com/beeware/toga/actions/runs/3642586521/jobs/6149928414) in today's log, it appears that the emulator's clock went backwards by two and a half minutes. My guess is that it started up with an inaccurate time, and then corrected itself from the internet.

**Comment by freakboy3742** [comment]
> > The only thing I can think of is maybe we should be running it through xvfb-run, as the backend unit tests do in tox.ini. Though if it was that, I would expect it to give a more useful error message.
> 
> Well, it turns out it was that. And I can reproduce the crash on my own Linux machine simply by setting the DISPLAY environment variable to empty to prevent the app from contacting the X server. That's a pretty terrible user experience, but I don't know which component is to blame.

Annoying, but at least it *works* once you know what the problem is, and it's something we can document as part of a CI testing guide for Briefcase.

**Review by freakboy3742 — APPROVED** [review]
Awesome - now we just need the actual tests :-)
